tests/test_co2eq_parameters.py

```diff
@@ -0,0 +1,188 @@
+            else:
+                modes.add(mode)
+    return modes
+
+
+def parse_json_file(path):
+    """Parse a JSON file."""
+    with open(path, encoding="utf-8") as file_:
+        return json.load(file_)
+
+
+class CO2eqParametersAll(unittest.TestCase):
+    """A test case for the CO2eq parameters."""
+    modes = get_possible_modes()
+    parameters = parse_json_file("config/co2eq_parameters_all.json")
+
+    @classmethod
+    def check_power_origin_ratios(cls, callback1, callback2):
+        """Apply the callback to each ratio in the 'powerOriginRatios' objects.
+        """
+        # It is useful to keep `test_power_origin_modes_are_valid` and
+        # `test_power_origin_ratios_sum_to_1` as separate tests, but they share
+        # a lot of identical code. This helper function helps avoid code
+        # duplication by factoring out the common logic and using callback
+        # functions to handle the differences.
+        fallback_zone_mixes = cls.parameters["fallbackZoneMixes"]
+        for zone, mixes in (("defaults", fallback_zone_mixes["defaults"]),
+                            *fallback_zone_mixes["zoneOverrides"].items()):
+            ratios = mixes["powerOriginRatios"]
+            if isinstance(ratios, list):
+                for ratio in ratios:
+                    callback1(ratio, zone)
+            else:
+                callback2(ratios, zone)
+
+    def test_power_origin_modes_are_valid(self):
+        """All modes in the 'powerOriginRatios' objects must be valid."""
+        def callback(ratio_, zone):
+            for mode, ratio in ratio_["value"].items():
+                if isinstance(ratio, numbers.Number):
+                    self.assertIn(
+                        mode,
+                        self.modes,
+                        msg=f"zone '{zone}' contains an invalid mode")
+
+        self.check_power_origin_ratios(callback, callback)
+
+    def test_power_origin_ratio_annual_lists_have_valid_dates(self):
+        """Lists of emission factors must include valid date strings."""
+        def callback(ratio, _zone):
+            # An exception will be raised if the string is not a valid
+            # datetime, thus failing the test.
+            datetime.datetime.fromisoformat(ratio["datetime"])
+
+        self.check_power_origin_ratios(callback, lambda ratio, zone: None)
+
+    def test_power_origin_ratios_sum_to_1(self):
+        """All ratios in the 'powerOriginRatios' objects must sum to
+        (approximately) 1.0.
+        """
+        def callback(ratio, zone):
+            self.assertAlmostEqual(
+                1.0,
+                sum(ratio for ratio in ratio["value"].values()
+                    if isinstance(ratio, numbers.Number)),
+                msg=f"zone '{zone}' ratios do not sum to (approximately) 1.0",
+                places=2)
+
+        self.check_power_origin_ratios(callback, callback)
+
+    def test_required_keys_are_present(self):
```

Comment:
I think the `test_required_keys_are_present` methods are overkill. It would probably be better to just remove them, and they'll exist implicitly in the other tests, relying on `KeyError`s to be raised if an expected key doesn't exist.